server/data/data-api/src/test/java/org/apache/james/rrt/lib/CanSendFromContract.java

```diff
@@ -103,50 +156,91 @@ default void allValidFromAddressesShouldContainOnlyUserAddressWhenUserHasNoAlias
 
         assertThat(canSendFrom().allValidFromAddressesForUser(USER))
             .containsExactlyInAnyOrder(USER.asMailAddress(), fromUser.asMailAddress());
     }
 
     @Test
     default void allValidFromAddressesShouldContainUserAddressAndAnAliasOfTheDomainUserFromAnotherDomain() throws Exception {
-        Username userAliasOtherDomain = Username.of(USER_ALIAS.getLocalPart() + "@" + OTHER_DOMAIN.asString());
+        Username userAliasOtherDomain = USER_ALIAS.withOtherDomain(Optional.of(OTHER_DOMAIN));
 
-        addDomainMapping(OTHER_DOMAIN, DOMAIN);
-        addAliasMapping(userAliasOtherDomain, USER);
+        redirectDomain(OTHER_DOMAIN).to(DOMAIN);
+        redirectUser(userAliasOtherDomain).to(USER);
 
-        Username userAliasMainDomain = Username.of(USER_ALIAS.getLocalPart() + "@" + DOMAIN.asString());
-        Username userOtherDomain = Username.of(USER.getLocalPart() + "@" + OTHER_DOMAIN.asString());
+        Username userAliasMainDomain = USER_ALIAS.withOtherDomain(Optional.of(DOMAIN));
+        Username userOtherDomain = USER.withOtherDomain(Optional.of(OTHER_DOMAIN));
         assertThat(canSendFrom().allValidFromAddressesForUser(USER))
             .containsExactlyInAnyOrder(USER.asMailAddress(), userAliasOtherDomain.asMailAddress(), userAliasMainDomain.asMailAddress(), userOtherDomain.asMailAddress());
     }
+
+    @Test
+    default void allValidFromAddressesShouldContainASendersAliasOfAnAliasOfTheUser() throws Exception {
+        Username userAliasBis = Username.of("aliasbis@" + DOMAIN.asString());
+        redirectUser(userAliasBis).to(USER_ALIAS);
+        redirectUser(USER_ALIAS).to(USER);
+
+        assertThat(canSendFrom().userCanSendFrom(USER, userAliasBis)).isTrue();
+    }
+
+    @Test
+    default void allValidFromAddressesShouldNotContainASendersAliasOfAnAliasOfAnAliasOfTheUser() throws Exception {
+        Username userAliasBis = Username.of("aliasbis@" + DOMAIN.asString());
+        Username userAliasTer = Username.of("aliaster@" + DOMAIN.asString());
+        redirectUser(userAliasTer).to(userAliasBis);
+        redirectUser(userAliasBis).to(USER_ALIAS);
+        redirectUser(USER_ALIAS).to(USER);
+
+        assertThat(canSendFrom().allValidFromAddressesForUser(USER))
+            .doesNotContain(userAliasTer.asMailAddress());
+    }
+
+    @Test
+    default void allValidFromAddressesShouldContainASendersAliasOfAnAliasInAnotherDomainOfTheUser() throws Exception {
+        Username userAlias = Username.of("aliasbis@" + OTHER_DOMAIN.asString());
+        Username userAliasBis = Username.of("aliaster@" + OTHER_DOMAIN.asString());
+        redirectUser(userAliasBis).to(userAlias);
+        redirectUser(userAlias).to(USER);
+
+        assertThat(canSendFrom().allValidFromAddressesForUser(USER))
+            .contains(userAliasBis.asMailAddress());
+    }
+
+    @Test
+    default void allValidFromAddressesShouldContainASendersAliasInAnotherDomainOfAnAliasOfTheUser() throws Exception {
```

Comment:
allValidFromAddressesShouldContainAnUserAliasFollowingADomainAliasResolution ?

It's hard to come up with good test names here...